README
nagios-plugins-linux-memory
===========================

This package contains two nagios plugins for respectivery checking memory and
swap usage on unix systems.

Usage:

        check_memory [-C] [-b,-k,-m,-g] [-w PERC] [-c PERC]
        check_swap [-b,-k,-m,-g] [-w PERC] [-c PERC]

        check_memory --help
        check_swap --help

Where:

 * -C, --caches:  count buffers and cached memory as free memory
 * -b,-k,-m,-g:   show output in bytes, KB (the default), MB, or GB
 * -w, --warning PERCENT:  warning threshold
 * -c, --critical PERCENT:  critical threshold

Examples:

        check_memory -C -m -w 80% -c 90%
        OK: 79.22% (810964 kB) used | mem_total=999MB, mem_used=791MB, mem_free=207MB, mem_shared=0MB, mem_buffers=1MB, mem_cached=190MB, mem_pageins=33803MB, mem_pageouts=18608MB
          # mem_total    : Total usable physical RAM
          # mem_used     : Total amount of physical RAM used by the system
          # mem_free     : Amount of RAM that is currently unused
          # mem_shared   : (Linux) Now always zero; not calculated
          # mem_buffers  : (Linux) Amount of physical RAM used for file buffers
          # mem_cached   : In-memory cache for files read from the disk (the page cache)
          # mem_pageins
          # mem_pageouts : (Linux) The number of memory pages the system has written in and out to disk
        
        check_swap -k -w 40% -c 60%
        WARNING: 42.70% (895104 kB) used | swap_total=2096444kB, swap_used=895104kB, swap_free=1201340kB, swap_cached=117024kB, swap_pageins=1593302kB, swap_pageouts=1281649kB
          # swap_total   : Total amount of swap space available
          # swap_used    : Total amount of swap used by the system
          # swap_free    : Amount of swap space that is currently unused
          # swap_cached  : (Linux) The amount of swap used as cache memory
          # swap_pageins 
          # swap_pageouts: (Linux) The number of swap pages the system has brought in and out


## Source code

The source code can be also found at
https://sites.google.com/site/davidemadrisan/opensource


## Installation

This package uses GNU autotools for configuration and installation.

If you have cloned the git repository then you will need to run
`autoreconf --install` to generate the required files.

Run `./configure --help` to see a list of available install options.
The plugin will be installed by default into `LIBEXECDIR`.

It is highly likely that you will want to customise this location to suit your
needs, i.e.:

        ./configure --libexecdir=/usr/lib/nagios/plugins

After `./configure` has completed successfully run `make install` and you're
done!


## Supported Platforms

This package is written in plain C, making as few assumptions as possible,
and sticking closely to ANSI C/POSIX.

This is a list of platforms this nagios plugin is known to compile and run on

        Linux with kernel 3.10 (openmamba milestone 2.90.0)
        Linux with kernel 2.6.32 (RHEL 6.X (Santiago))
        Linux with kernel 2.6.18 (RHEL 5.X (Tikanga))
	OpenBSD 5.0 

My special thanks to grex.org team for giving me a free OpenBSD shell account


## Bugs

If you find a bug please create an issue in the project bug tracker at
https://github.com/madrisan/nagios-plugins-memory/issues



NEWS

    Major changes betwhhn "nagios-plugins-memory" version 2 and 3

* The code has been splitted into two different binary programs:
  check_memory and check_swap.
* configure: new option '--with-proc-meminfo=PATH'.
* check_swap: display on linux the number of swapins and swapouts,
  in kilobytes (since the last boot), and the amount of swap used as
  cache memory.
* check_memory: add on linux 'mem_pageins' and 'mem_pageouts'.
* These plugins can now be compiled by the 'clang' compiler without
  any warning message.

======================================================================

    Major changes betwhhn "nagios-plugins-memory" version 1 and 2

* A new architecture is supported: OpenBSD (tested on v5.0).
* The plugin has been renamed by removing the '-linux' string.
* The os-dependent code has been splitted into different files.
* Enable some extra compiler warnings and fix them all.

======================================================================

    Major changes in "nagios-plugins-linux-memory" version 1

* First release.

======================================================================


nputils.c
/*
 * License: GPL
 * Copyright (c) 2006 Nagios Plugins Development Team
 *
 * Library of useful functions for nagios plugins
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

#include <string.h>
#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif

#include "nputils.h"

/*
 * Returns TRUE if alert should be raised based on the range 
 */
int
check_range (double value, range * my_range)
{
  int no = FALSE;
  int yes = TRUE;

  if (my_range->alert_on == INSIDE)
    {
      no = TRUE;
      yes = FALSE;
    }

  if (my_range->end_infinity == FALSE && my_range->start_infinity == FALSE)
    {
      if ((my_range->start <= value) && (value <= my_range->end))
	return no;
      else
	return yes;
    }
  else if (my_range->start_infinity == FALSE
	   && my_range->end_infinity == TRUE)
    {
      if (my_range->start <= value)
	return no;
      else
	return yes;
    }
  else if (my_range->start_infinity == TRUE
	   && my_range->end_infinity == FALSE)
    {
      if (value <= my_range->end)
	return no;
      else
	return yes;
    }
  else
    return no;
}

int
get_status (double value, thresholds * my_thresholds)
{
  if (my_thresholds->critical != NULL)
    {
      if (check_range (value, my_thresholds->critical) == TRUE)
	return STATE_CRITICAL;
    }
  if (my_thresholds->warning != NULL)
    {
      if (check_range (value, my_thresholds->warning) == TRUE)
	return STATE_WARNING;
    }
  return STATE_OK;
}

void
set_range_start (range * this, double value)
{
  this->start = value;
  this->start_infinity = FALSE;
}

void
set_range_end (range * this, double value)
{
  this->end = value;
  this->end_infinity = FALSE;
}

range *
parse_range_string (char *str)
{
  range *temp_range;
  double start;
  double end;
  char *end_str;

  temp_range = (range *) malloc (sizeof (range));

  /*
   * Set defaults 
   */
  temp_range->start = 0;
  temp_range->start_infinity = FALSE;
  temp_range->end = 0;
  temp_range->end_infinity = TRUE;
  temp_range->alert_on = OUTSIDE;

  if (str[0] == '@')
    {
      temp_range->alert_on = INSIDE;
      str++;
    }

  end_str = index (str, ':');
  if (end_str != NULL)
    {
      if (str[0] == '~')
	temp_range->start_infinity = TRUE;
      else
	{
	  start = strtod (str, NULL);	/* Will stop at the ':' */
	  set_range_start (temp_range, start);
	}
      end_str++;		/* Move past the ':' */
    }
  else
    {
      end_str = str;
    }
  end = strtod (end_str, NULL);
  if (strcmp (end_str, "") != 0)
    set_range_end (temp_range, end);

  if (temp_range->start_infinity == TRUE ||
      temp_range->end_infinity == TRUE ||
      temp_range->start <= temp_range->end)
    {
      return temp_range;
    }
  free (temp_range);
  return NULL;
}

/*
 * returns 0 if okay, otherwise 1 
 */
int
set_thresholds (thresholds ** my_thresholds, char *warn_string,
		char *critical_string)
{
  thresholds *temp_thresholds = NULL;

  if ((temp_thresholds = malloc (sizeof (thresholds))) == NULL)
    {
      printf ("Cannot allocate memory: %s", strerror (errno));
      exit (STATE_UNKNOWN);
    }

  temp_thresholds->warning = NULL;
  temp_thresholds->critical = NULL;

  if (warn_string != NULL)
    {
      if ((temp_thresholds->warning =
	   parse_range_string (warn_string)) == NULL)
	{
	  return NP_RANGE_UNPARSEABLE;
	}
    }
  if (critical_string != NULL)
    {
      if ((temp_thresholds->critical =
	   parse_range_string (critical_string)) == NULL)
	{
	  return NP_RANGE_UNPARSEABLE;
	}
    }

  *my_thresholds = temp_thresholds;

  return 0;
}

const char *
state_text (int result)
{
  switch (result)
    {
    case STATE_OK:
      return "OK";
    case STATE_WARNING:
      return "WARNING";
    case STATE_CRITICAL:
      return "CRITICAL";
    case STATE_DEPENDENT:
      return "DEPENDENT";
    default:
      return "UNKNOWN";
    }
}


nputils.h
#pragma once

#ifndef TRUE
# define TRUE 1
#endif

#ifndef FALSE
# define FALSE 0
#endif

/* Return codes for _set_thresholds */
#define NP_RANGE_UNPARSEABLE 1
#define NP_WARN_WITHIN_CRIT  2

#define OUTSIDE 0
#define INSIDE  1

enum
{
  STATE_OK,
  STATE_WARNING,
  STATE_CRITICAL,
  STATE_UNKNOWN,
  STATE_DEPENDENT
};

/* see: nagios-plugins-1.4.15/lib/utils_base.h */
typedef struct range_struct
{
  double start;
  int start_infinity;		/* FALSE (default) or TRUE */
  double end;
  int end_infinity;
  int alert_on;			/* OUTSIDE (default) or INSIDE */
} range;

typedef struct thresholds_struct
{
  range *warning;
  range *critical;
} thresholds;

int get_status (double, thresholds *);
int set_thresholds (thresholds **, char *, char *);
const char *state_text (int);
void die (int, const char *, ...)
        __ATTRIBUTE_NORETURN__
        __attribute__ ((__format__ (__printf__, 2, 3)));



check_swap.c
/*
 * License: GPLv2
 * Copyright (c) 2014 Davide Madrisan <davide.madrisan@gmail.com>
 *
 * A Nagios plugin to check the swap usage on unix.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * This software is based on the source code of the tool "free" (procps 3.2.8)
 */

#include "config.h"

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "nputils.h"
#include "meminfo.h"

static const char *program_name = "check_swap";
static const char *program_version = PACKAGE_VERSION;
static const char *program_copyright =
  "Copyright (C) 2014 Davide Madrisan <" PACKAGE_BUGREPORT ">";

static void __ATTRIBUTE_NORETURN__ usage (FILE * out)
{
  fprintf (out,
           "%s, version %s - check swap usage.\n",
           program_name, program_version);
  fprintf (out, "%s\n\n", program_copyright);
  fprintf (out,
           "Usage: %s [-b,-k,-m,-g] -w PERC -c PERC\n",
           program_name);
  fprintf (out, "       %s -h\n", program_name);
  fprintf (out, "       %s -V\n\n", program_name);
  fputs ("\
Options:\n\
  -b,-k,-m,-g      show output in bytes, KB (the default), MB, or GB\n\
  -w, --warning PERCENT   warning threshold\n\
  -c, --critical PERCENT   critical threshold\n\
  -h, --help       display this help and exit\n\
  -v, --version    output version information and exit\n\n", out);
  fprintf (out, "\
Examples:\n\
  %s -w 30%% -c 50%%\n\n", program_name);

  exit (out == stderr ? STATE_UNKNOWN : STATE_OK);
}

static void __ATTRIBUTE_NORETURN__ print_version (void)
{
  printf ("%s, version %s\n", program_name, program_version);
  printf ("%s\n", program_copyright);
  fputs ("\
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>\n\n\
This is free software; you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n", stdout);

  exit (STATE_OK);
}

static struct option const longopts[] = {
  {(char *) "critical", required_argument, NULL, 'c'},
  {(char *) "warning", required_argument, NULL, 'w'},
  {(char *) "byte", no_argument, NULL, 'b'},
  {(char *) "kilobyte", no_argument, NULL, 'k'},
  {(char *) "megabyte", no_argument, NULL, 'm'},
  {(char *) "gigabyte", no_argument, NULL, 'g'},
  {(char *) "help", no_argument, NULL, 'h'},
  {(char *) "version", no_argument, NULL, 'V'},
  {NULL, 0, NULL, 0}
};

int
main (int argc, char **argv)
{
  int c, status;
  int shift = 10;
  char *critical = NULL, *warning = NULL;
  char *units = NULL;
  char *status_msg;
  char *perfdata_msg;
  thresholds *my_threshold = NULL;
  float percent_used = 0;

  while ((c = getopt_long (argc, argv, "c:w:bkmghV", longopts, NULL)) != -1)
    {
      switch (c)
        {
        default:
          usage (stderr);
        case 'c':
          critical = optarg;
          break;
        case 'w':
          warning = optarg;
          break;
        case 'h':
          usage (stdout);
        case 'V':
          print_version ();
        case 'b': shift = 0;  units = strdup ("B"); break;
        case 'k': shift = 10; units = strdup ("kB"); break;
        case 'm': shift = 20; units = strdup ("MB"); break;
        case 'g': shift = 30; units = strdup ("GB"); break;
        }
    }

  status = set_thresholds (&my_threshold, warning, critical);
  if (status == NP_RANGE_UNPARSEABLE)
    usage (stderr);

  /* output in kilobytes by default */
  if (units == NULL)
    units = strdup ("kB");

  swapinfo ();

  if (kb_swap_total != 0)
    percent_used = (kb_swap_used * 100.0 / kb_swap_total);

  status = get_status (percent_used, my_threshold);
  free (my_threshold);

  status_msg = get_swap_status (status, percent_used, shift, units);
  perfdata_msg = get_swap_perfdata (shift, units);

  printf ("%s | %s\n", status_msg, perfdata_msg);

  free (units);
  free (status_msg);
  free (perfdata_msg);

  return status;
}


meminfo-openbsd.c
/*
 * License: GPLv2
 * Copyright (c) 2014 Davide Madrisan <davide.madrisan@gmail.com>
 *
 * A Nagios plugin to check memory and swap usage on openbsd.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * This software is based on the source code of the tool "free" (procps 3.2.8)
 */

#include "config.h"

#ifndef _GNU_SOURCE
#define _GNU_SOURCE /* activate extra prototypes for glibc */
#endif

#include <sys/param.h>
#include <sys/types.h>
#include <sys/mount.h>
#include <sys/sysctl.h>
#if HAVE_SYS_SWAP_H
# include <sys/swap.h>
#endif
#include <unistd.h>    /* getpagesize */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "nputils.h"

#define SU(X) ( ((unsigned int)(X) << 10) >> shift ), units

# define NUM_AVERAGES    3
  /* Log base 2 of 1024 is 10 (2^10 == 1024) */
# define LOG1024         10

/* these are for getting the memory statistics */
static int      pageshift;      /* log base 2 of the pagesize */

/* define pagetok in terms of pageshift */
#define pagetok(size) ((size) << pageshift)

int kb_main_used;
int kb_main_cached;
int kb_main_free;
int kb_main_total;

int kb_swap_used;
int kb_swap_free;
int kb_swap_total;

int
get_system_pagesize (void)
{
  int pagesize;

  /*
   * get the page size with "getpagesize" and calculate pageshift from
   * it
   * fixme: we should use sysconf(_SC_PAGESIZE) instead
   */
  pagesize = getpagesize ();
  pageshift = 0;
  while (pagesize > 1)
    {
      pageshift++;
      pagesize >>= 1;
    }

  /* we only need the amount of log(2)1024 for our conversion */
  pageshift -= LOG1024;

  return (0);
}

int
swapmode (int *used, int *total)
{
  struct swapent *swdev;
  int nswap, rnswap, i;

  nswap = swapctl (SWAP_NSWAP, 0, 0);
  if (nswap == 0)
    return 0;

  swdev = calloc (nswap, sizeof (*swdev));
  if (swdev == NULL)
    return 0;

  rnswap = swapctl (SWAP_STATS, swdev, nswap);
  if (rnswap == -1)
    {
      free (swdev);
      return 0;
    }

  /* if rnswap != nswap, then what? */

  /* Total things up */
  *total = *used = 0;
  for (i = 0; i < nswap; i++)
    {
      if (swdev[i].se_flags & SWF_ENABLE)
        {
          *used += (swdev[i].se_inuse / (1024 / DEV_BSIZE));
          *total += (swdev[i].se_nblks / (1024 / DEV_BSIZE));
        }
    }
  free (swdev);
  return 1;
}

void
meminfo (int cache_is_free)
{
  static int vmtotal_mib[] = { CTL_VM, VM_METER };
  static int bcstats_mib[] = { CTL_VFS, VFS_GENERIC, VFS_BCACHESTAT };
  struct vmtotal vmtotal;
  struct bcachestats bcstats;
  size_t size;

  if (get_system_pagesize() == -1)
    die (STATE_UNKNOWN, "RUNTIME ERROR: get_system_pagesize failed\n");

  /* get total -- systemwide main memory usage structure */
  size = sizeof (vmtotal);
  if (sysctl (vmtotal_mib, 2, &vmtotal, &size, NULL, 0) < 0)
    { 
      bzero (&vmtotal, sizeof (vmtotal));
      die (STATE_UNKNOWN, "RUNTIME ERROR: sysctl failed\n");
    } 
  size = sizeof (bcstats);
  if (sysctl (bcstats_mib, 3, &bcstats, &size, NULL, 0) < 0)
    { 
      bzero (&bcstats, sizeof (bcstats));
      die (STATE_UNKNOWN, "RUNTIME ERROR: sysctl failed\n");
    } 

  /* convert memory stats to Kbytes */
  kb_main_total = pagetok (vmtotal.t_rm);
  kb_main_used = pagetok (vmtotal.t_arm);
  kb_main_free = pagetok (vmtotal.t_free);;
  kb_main_cached = pagetok (bcstats.numbufpages);

  if (cache_is_free)
    {
      kb_main_used -= kb_main_cached;
      kb_main_free += kb_main_cached;
    }
}

void
swapinfo (void)
{
  static int bcstats_mib[] = { CTL_VFS, VFS_GENERIC, VFS_BCACHESTAT };
  struct bcachestats bcstats;
  size_t size;

  if (get_system_pagesize() == -1)
    die (STATE_UNKNOWN, "RUNTIME ERROR: get_system_pagesize failed\n");

  size = sizeof (bcstats);
  if (sysctl (bcstats_mib, 3, &bcstats, &size, NULL, 0) < 0)
    { 
      bzero (&bcstats, sizeof (bcstats));
      die (STATE_UNKNOWN, "RUNTIME ERROR: sysctl failed\n");
    } 

  if (!swapmode (&kb_swap_used, &kb_swap_total))
    {
      kb_swap_total = 0;
      kb_swap_used = 0;
    }
  kb_swap_free = kb_swap_total - kb_swap_used;
}

char *
get_memory_status (int status, float percent_used, int shift,
                   const char *units)
{
  char *msg;
  int ret;

  ret = asprintf (&msg, "%s: %.2f%% (%d kB) used", state_text (status),
                  percent_used, kb_main_used);

  if (ret < 0)
    die (STATE_UNKNOWN, "Error getting memory status\n");
  
  return msg;
}

char *
get_swap_status (int status, float percent_used, int shift,
                 const char *units)
{
  char *msg;
  int ret;

  ret = asprintf (&msg, "%s: %.2f%% (%d kB) used", state_text (status),
                  percent_used, kb_swap_used);

  if (ret < 0)
    die (STATE_UNKNOWN, "Error getting swap status\n");

  return msg;
}

char *
get_memory_perfdata (int shift, const char *units)
{
  char *msg;
  int ret;

  ret = asprintf (&msg,
                  "mem_total=%d%s, "
                  "mem_used=%d%s, "
                  "mem_free=%d%s, "
                  "mem_cached=%d%s\n",
                  SU (kb_main_total),
                  SU (kb_main_used),
                  SU (kb_main_free),
                  SU (kb_main_cached));

  if (ret < 0)
    die (STATE_UNKNOWN, "Error getting memory perfdata\n");

  return msg;
}

char *
get_swap_perfdata (int shift, const char *units)
{
  char *msg;
  int ret;

  ret = asprintf (&msg,
                  "swap_total=%d%s, "
                  "swap_used=%d%s, "
                  "swap_free=%d%s\n",
                  SU (kb_swap_total),
                  SU (kb_swap_used),
                  SU (kb_swap_free));

  if (ret < 0)
    die (STATE_UNKNOWN, "Error getting swap perfdata\n");

  return msg;
}


check_memory.c
/*
 * License: GPLv2
 * Copyright (c) 2014 Davide Madrisan <davide.madrisan@gmail.com>
 *
 * A Nagios plugin to check memory usage on unix.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * This software is based on the source code of the tool "free" (procps 3.2.8)
 */

#include "config.h"

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "nputils.h"
#include "meminfo.h"

static const char *program_name = "check_memory";
static const char *program_version = PACKAGE_VERSION;
static const char *program_copyright =
  "Copyright (C) 2014 Davide Madrisan <" PACKAGE_BUGREPORT ">";

static void __ATTRIBUTE_NORETURN__ usage (FILE * out)
{
  fprintf (out,
           "%s, version %s - check memory usage.\n",
           program_name, program_version);
  fprintf (out, "%s\n\n", program_copyright);
  fprintf (out,
           "Usage: %s [-b,-k,-m,-g] [-C] -w PERC -c PERC\n",
           program_name);
  fprintf (out, "       %s -h\n", program_name);
  fprintf (out, "       %s -V\n\n", program_name);
  fputs ("\
Options:\n\
  -b,-k,-m,-g      show output in bytes, KB (the default), MB, or GB\n\
  -C, --caches     count buffers and cached memory as free memory\n\
  -w, --warning PERCENT   warning threshold\n\
  -c, --critical PERCENT   critical threshold\n\
  -h, --help       display this help and exit\n\
  -v, --version    output version information and exit\n\n", out);
  fprintf (out, "\
Examples:\n\
  %s -C -w 80%% -c90%%\n", program_name);

  exit (out == stderr ? STATE_UNKNOWN : STATE_OK);
}

static void __ATTRIBUTE_NORETURN__ print_version (void)
{
  printf ("%s, version %s\n", program_name, program_version);
  printf ("%s\n", program_copyright);
  fputs ("\
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>\n\n\
This is free software; you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n", stdout);

  exit (STATE_OK);
}

static struct option const longopts[] = {
  {(char *) "caches", no_argument, NULL, 'C'},
  {(char *) "critical", required_argument, NULL, 'c'},
  {(char *) "warning", required_argument, NULL, 'w'},
  {(char *) "byte", no_argument, NULL, 'b'},
  {(char *) "kilobyte", no_argument, NULL, 'k'},
  {(char *) "megabyte", no_argument, NULL, 'm'},
  {(char *) "gigabyte", no_argument, NULL, 'g'},
  {(char *) "help", no_argument, NULL, 'h'},
  {(char *) "version", no_argument, NULL, 'V'},
  {NULL, 0, NULL, 0}
};

int
main (int argc, char **argv)
{
  int c, status;
  int cache_is_free = 0;
  int shift = 10;
  char *critical = NULL, *warning = NULL;
  char *units = NULL;
  char *status_msg;
  char *perfdata_msg;
  thresholds *my_threshold = NULL;
  float percent_used = 0;

  while ((c = getopt_long (argc, argv, "MSCc:w:bkmghV", longopts, NULL)) != -1)
    {
      switch (c)
        {
        default:
          usage (stderr);
        case 'C':
          cache_is_free = 1;
          break;
        case 'c':
          critical = optarg;
          break;
        case 'w':
          warning = optarg;
          break;
        case 'h':
          usage (stdout);
        case 'V':
          print_version ();
        case 'b': shift = 0;  units = strdup ("B"); break;
        case 'k': shift = 10; units = strdup ("kB"); break;
        case 'm': shift = 20; units = strdup ("MB"); break;
        case 'g': shift = 30; units = strdup ("GB"); break;
        }
    }

  status = set_thresholds (&my_threshold, warning, critical);
  if (status == NP_RANGE_UNPARSEABLE)
    usage (stderr);

  /* output in kilobytes by default */
  if (units == NULL)
    units = strdup ("kB");

  meminfo (cache_is_free);

  if (kb_main_total != 0)
    percent_used = (kb_main_used * 100.0 / kb_main_total);

  status = get_status (percent_used, my_threshold);
  free (my_threshold);

  status_msg = get_memory_status (status, percent_used, shift, units);
  perfdata_msg = get_memory_perfdata (shift, units);

  printf ("%s | %s\n", status_msg, perfdata_msg);

  free (units);
  free (status_msg);
  free (perfdata_msg);

  return status;
}
